RefApp/ViewControllers/LoginViewController.swift

```diff
@@ -0,0 +1,173 @@
+            
+            loginButton.anchor(top: passwordTextField.bottomAnchor, bottom: nil, leading: view.safeAreaLayoutGuide.leadingAnchor, trailing: view.safeAreaLayoutGuide.trailingAnchor, padding: .init(top:20, left: 16, bottom:0, right: -16))
+            
+        } else {
+            topHolderView.anchor(top: view.topAnchor, bottom: bottomHolderView.topAnchor, leading: view.leadingAnchor, trailing: view.trailingAnchor)
+            
+            bottomHolderView.anchor(top: topHolderView.bottomAnchor, bottom: nil, leading: view.leadingAnchor, trailing: view.trailingAnchor)
+            
+            usernameTextField.anchor(top: bottomHolderView.topAnchor, bottom: nil, leading: view.leadingAnchor, trailing: view.trailingAnchor, padding: .init(top:20, left: 16, bottom:0, right: -16))
+            
+            passwordTextField.anchor(top: usernameTextField.bottomAnchor, bottom: nil, leading: view.leadingAnchor, trailing: view.trailingAnchor, padding: .init(top:20, left: 16, bottom:0, right: -16))
+            
+            loginButton.anchor(top: passwordTextField.bottomAnchor, bottom: nil, leading: view.leadingAnchor, trailing: view.trailingAnchor, padding: .init(top:20, left: 16, bottom:0, right: -16))
+        }
+        
+        topHolderView.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 1/3).isActive = true
+        
+        logoImageView.anchor(top: nil, bottom: topHolderView.bottomAnchor, leading: nil, trailing: nil, padding: .init(top: 0, left: 0, bottom: 0, right: 0), size: .init(width: 0, height: 0))
+        logoImageView.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
+        
+        bottomHolderView.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 2/3).isActive = true
+    }
+}
+
+
+
+// MARK: - Actions
+extension LoginViewController {
+    
+    /// Authenticate the user with Authentication Endpoint
+    @objc fileprivate func authenticateUser() {
+        
+        view.endEditing(true)
+        
+        let okAction = UIAlertAction(title: NSLocalizedString("Ok", comment: ""), style: .cancel, handler: {
+            (alert: UIAlertAction!) -> Void in
+        })
+        
+        guard let username = usernameTextField.text, let password = passwordTextField.text, username != "", password != "" else {
+            
+            self.popupAlert(title:NSLocalizedString("Sorry", comment: "") , message: NSLocalizedString("Please fill all fields", comment: ""), actions: [okAction], preferedStyle: .alert)
+            
+            return
+        }
+        
+        Authenticate(environment: environment)
+            .login(username: username, password: password)
+            .request()
+            .validate()
+            .response{ [weak self] in
+                if let error = $0.error {
+                    self?.popupAlert(title: error.message , message: error.localizedDescription, actions: [okAction], preferedStyle: .alert)
```

Comment:
ExposureError provides a lot more details (no exposed through localizedDescription) which could be helpful in this context.

In previous implementations we have used something along the lines of:

```Swift
let message = "\(error.code) " + error.message + "\n" + (error.info ?? "")
self?.popupAlert(title: error.domain , message: message, actions: [okAction], preferedStyle: .alert)
```

Reply:
nice, seems more informative. I will add that.